pkg/vere/disk.c

```diff
@@ -867,66 +874,370 @@ u3_disk_init(c3_c* pax_c, u3_disk_cb cb_u)
+    c3_d lat_d;
+    if ( c3n == u3_disk_epoc_last(log_u, &lat_d) ) {
+      fprintf(stderr, "disk: failed to load epoch number\r\n");
+      c3_free(log_u);
+      return 0;
+    }
+
+    //  set path to latest epoch
+    c3_c epo_c[8193];
+    snprintf(epo_c, 8192, "%s/0i%" PRIu64, log_c, lat_d);
+
+    //  get vere version from the latest epoch
+    c3_c fil_c[8193];
+    snprintf(fil_c, 8192, "%s/vere.txt", epo_c);
+    if ( 0 != access(fil_c, F_OK) ) {
+      fprintf(stderr, "disk: failed to access %s\r\n", fil_c);
+      c3_free(log_u);
+      return 0;
+    }
+
+    //  read string from vere.txt file
+    c3_c ver_c[8193];
     {
-      const size_t siz_i =
-      // 500 GiB is as large as musl on aarch64 wants to allow
-      #if (defined(U3_CPU_aarch64) && defined(U3_OS_linux))
-        0x7d00000000;
-      #else
-        0x10000000000;
-      #endif
-
-      if ( 0 == (log_u->mdb_u = u3_lmdb_init(log_c, siz_i)) ) {
-        fprintf(stderr, "disk: failed to initialize database\r\n");
-        c3_free(log_c);
+      FILE* fil_u = fopen(fil_c, "r");
+      if ( !fil_u ) {
+        fprintf(stderr, "disk: failed to open %s\r\n", fil_c);
         c3_free(log_u);
         return 0;
       }
+      if ( !fgets(ver_c, 8192, fil_u) ) {
+        fprintf(stderr, "disk: failed to read %s\r\n", fil_c);
+        c3_free(log_u);
+        return 0;
+      }
+      fclose(fil_u);
     }
 
-    c3_free(log_c);
-  }
+    //  if binary version from vere.txt is different 
+    //  than current version, then create a new epoch
+    if ( 0 != strcmp(ver_c, URBIT_VERSION) ) {
+      fprintf(stderr, "disk: binary version mismatch, creating new epoch\r\n");
+      if ( c3n == u3_disk_epoc_init(log_u) ) {
+        fprintf(stderr, "disk: failed to initialize epoch\r\n");
+        c3_free(log_u);
+        return 0;
+      }
+    }
 
-  //  get the latest event number from the db
-  //
-  {
-    log_u->dun_d = 0;
-    c3_d fir_d;
+    //  initialize epoch's db
+    {
 
-    if ( c3n == u3_lmdb_gulf(log_u->mdb_u, &fir_d, &log_u->dun_d) ) {
-      fprintf(stderr, "disk: failed to load latest event from database\r\n");
-      c3_free(log_u);
+      if ( 0 == (log_u->mdb_u = u3_lmdb_init(epo_c)) ) {
+        fprintf(stderr, "disk: failed to initialize database\r\n");
+        c3_free(log_u);
+        return 0;
+      }
+    }
+    fprintf(stderr, "disk: loaded epoch 0i%" PRIu64 "\r\n", lat_d);
+
+    //  get first/last event numbers from lmdb
+    c3_d fir_d, las_d;
+    if ( c3n == u3_lmdb_gulf(log_u->mdb_u, &fir_d, &las_d) ) {
+      fprintf(stderr, "disk: failed to get first/last event numbers\r\n");
       return 0;
     }
 
+    //  initialize dun_d/sen_d values
+    if ( 0 == las_d ) {               //  fresh epoch (no events in lmdb yet)
+      if ( 0 == lat_d ) {             //  first epoch
+        log_u->dun_d = 0;
+      } else {                        //  not first epoch
+        log_u->dun_d = lat_d - 1;     //  set dun_d to last event in prev epoch
```

Comment:
This looks wrong. `lat_d` is the previous epoch number, right? What we need is the number of the last event in the previous epoch.

Reply:
Apologies for lack of clarity. `lat_d` is the latest epoch number *after* creating a new one in the migrate function. Therefore it will be an empty epoch and named N+1 (where N is the last event of the previous epoch) or 0 (if it's the first epoch). Hope that makes sense. The code does indeed work.

Reply:
Just realized this isn't reflecting the design though... According to our call this past week, we want the epoch numbers to be named N (not N+1) where N is the last event of the previous epoch. Will fix.